### DISCUSSION TITLE
How to implement an mDNS responder?

### DISCUSSION THREAD

**Discussion by u77345** [rsta2/circle#427] [Development]
I wonder what's the best way of implementing an mDNS responder. In line with the design patterns of Circle. Is there a similar component example to follow?

**Comment by rsta2** [comment]
It's probably not possible to implement mDNS with the unmodified network stack 
of Circle, because it does not support multicasts.

**Reply by u77345** [reply]
Thinking only transmitting. Would the stack prevent using the 224.0.0.251 destination?

**Reply by rsta2** [reply]
I suppose, it sends an ARP request for each non-broadcast IP address and waits for someone to respond.

**Comment by probonopd** [comment]
I'd also be interested in this. Could be used, e.g, for [RTP MIDI](https://github.com/dwhinham/mt32-pi/wiki/Networking%3A-RTP-MIDI-%28AppleMIDI%29) (AppleMIDI) announcement on the network for plug-and-play.

**Reply by rsta2** [reply]
I'm currently not sure, if this requires multicast transmit only, or if multicast receive is required. Multicast receive would be much more difficult, because it needs support from the various network interface drivers, which is not available at the moment. But I put it on my TODO list.

**Comment by rsta2** [comment]
There is IP multicast send support (conformity level 1 according to RFC1112) on the *develop* branch. This could be enough to publish an Apple-MIDI service using Bonjour (mDNS-SD) on an Ethernet network. IP multicast receive (level 2) would require much more effort, because it requires four Ethernet NIC drivers updated for multicast receive (including multicast filter support) and the implementation of the IGMP group management protocol, which has three different versions. There would be much reading required to do so. That's why I hope, it will work without this.

The attached test program publishes an `_apple-midi._udp` service, sleeps for 60 seconds and deletes the service again. I was only able to test this with a Linux host running `avahi-browse`, which should display, while the program is running:

```
$ avahi-browse -r _apple-midi._udp
+ enp1s0 IPv4 midi-test                                     _apple-midi._udp     local
= enp1s0 IPv4 midi-test                                     _apple-midi._udp     local
   hostname = [raspberrypi.local]
   address = [192.168.0.146]
   port = [10000]
   txt = ["Sample service"]
- enp1s0 IPv4 midi-test                                     _apple-midi._udp     local
^CGot SIGINT, quitting.
$
```
I needed to open the firewall on my Linux PC, so that it works.

Unfortunately I do not know yet, if this works with macOS and/or Windows. This has to be tested. If so, the class `CmDNSPublisher` in the test program could be added to the *net* library of Circle.

Of course, this is not an implementation of RTP-MIDI. It only publishes a RTP-MIDI service for Plug-and-Play configuration, the service itself must be provided by yourself.

[mdns-publisher.zip](https://github.com/user-attachments/files/15685043/mdns-publisher.zip)

**Reply by u77345** [reply]
Thanks Rene! The basic service publishing and removal works and recognized on a Mac. I'll have more details about multiple services and multiple txt records.

**Reply by rsta2** [reply]
@u77345 Thanks for info! Just let me know, if you have suggestions for improving the `CmDNSPublisher` class. My intention was, that for multiple services this class is instantiated multiple times. I was not aware, that multiple TXT records are allowed, but I understand, that this can be the case.

**Comment by rsta2** [comment]
An updated class `CmDNSResponder` is on the *develop* branch now with a test in *test/mdns-publisher*. The class now supports multiple services in one instance of the class and multiple TXT entries.

**Comment by omersiar** [comment]
I'm trying to implement this on MiniDexed, so far it works when it is first boots up then it stops responding to discovery.
Also thinking a way around, publishing and unpublishing at intervals.

Arduino mDNS library has [mdns.run() ](https://github.com/arduino-libraries/ArduinoMDNS/blob/master/examples/WiFi/WiFiRegisteringServicesWithTxtRecord/WiFiRegisteringServicesWithTxtRecord.ino#L110C3-L110C14)

I tried to understand recommended intervals for mdns requests but it seems it does not work as intervals.
What would be best to keep responding to discovery requests?

**Reply by rsta2** [reply]
The mDNS client in Circle is a send-only client. It cannot respond to requests. This is caused by the missing multicast receive support and cannot be changed easily. The mDNS client repeats the publish message in an [120 seconds](https://github.com/rsta2/circle/blob/c243194670720c8a13e6a553654b6d26ec7845a0/include/circle/net/mdnspublisher.h#L66) interval. If this interval is too long, you could decrease this value.

**Reply by omersiar** [reply]
Oh I see.  That helps.
